src/cats/CatSlideshow.js

```diff
@@ -0,0 +1,219 @@
+import { useEffect, useRef, useReducer } from "react";
+import { preloadImage } from "./utils.js";
+import getCatsApiFetchParams from "./getCatsApiFetchParams.js";
+import usePaginatedFetch from "./usePaginatedFetch";
+import CatSlideshowControls from "./CatSlideshowControls.js";
+import { SlideAnimation, Direction } from "./SlideAnimation.js";
+import { LoadingCard } from "./Cards.js";
+import "./CatSlideshow.css";
+
+const PAGE_SIZE = 20;
+const PAGE_PREFETCH = 8;
+const MAX_PREFETCH = 2;
+
+function fetchPageFromCatsApi(pageIndex, selectedBreedID, fetchPage) {
+  const { url, options } = getCatsApiFetchParams("/images/search", {
+    breed_id: selectedBreedID === "all" ? "" : selectedBreedID,
+    limit: PAGE_SIZE,
+    order: "ASC",
+    page: pageIndex,
+  });
+
+  const getPageData = async (response) => {
+    const json = await response.json();
+    return json.map((item) => ({ id: item.id, url: item.url }));
+  };
+
+  const getMetadata = (response) => {
+    return Object.fromEntries(
+      [...response.headers.entries()].filter(
+        (entry) => entry[0] === "pagination-count"
+      )
+    );
+  };
+
+  fetchPage(url, options, pageIndex, selectedBreedID, getPageData, getMetadata);
+}
+
+function getPageIndex(index) {
+  return Math.floor(index / PAGE_SIZE);
+}
+
+function getImageURL(pages, index) {
+  const pageIndex = getPageIndex(index);
+  const offset = index % PAGE_SIZE;
+  return pages?.[pageIndex]?.data?.[offset]?.url;
+}
+
+function initCatSlideshowState() {
+  return {
+    direction: Direction.Next,
+    index: 0,
+    maxIndex: null,
+    visibleIndex: null,
+    readyMap: {},
+  };
+}
+
+function catSlideshowReducer(state, action) {
+  //console.log("%ccat slideshow: " + action.type, "color: red", action, state);
+
+  function changeIndex(state, increment) {
+    const newIndex = Math.min(
+      Math.max(0, state.index + increment),
+      state.maxIndex
+    );
+    const isReady = state.indexReadyMap[newIndex] === "ready";
+
+    return {
+      ...state,
+      direction: increment > 0 ? Direction.Next : Direction.Previous,
+      index: newIndex,
+      visibleIndex: isReady ? newIndex : state.visibleIndex,
```

Comment:
Ah this is neat, so you keep showing the previous one until the new one loads

Reply:
yes. I like how this turned out : )